core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -727,40 +727,44 @@ private[master] class Master(
 
   def removeApplication(app: ApplicationInfo, state: ApplicationState.Value) {
     if (apps.contains(app)) {
-      logInfo("Removing app " + app.id)
-      apps -= app
-      idToApp -= app.id
-      endpointToApp -= app.driver
-      addressToApp -= app.driver.address
-      if (completedApps.size >= RETAINED_APPLICATIONS) {
-        val toRemove = math.max(RETAINED_APPLICATIONS / 10, 1)
-        completedApps.take(toRemove).foreach( a => {
-          appIdToUI.remove(a.id).foreach { ui => webUi.detachSparkUI(ui) }
-          applicationMetricsSystem.removeSource(a.appSource)
-        })
-        completedApps.trimStart(toRemove)
-      }
-      completedApps += app // Remember it in our history
-      waitingApps -= app
+      synchronized{
```

Comment:
I've added duplicate condition to avoid synchronization in case `app` doesn't exist in `apps`.if that is not required , I'll remove double checking.

Yes, `removeApplication` can be called concurrently with the same `app` instance from message loop in  Master.scala and also from handleAppKillRequest  in  MasterPage.scala.
This fix is not for handling concurrent modification of 'app'